File: Java/src/model/Statements/HeapAllocation.java
package model.Statements;

import Exceptions.DivideByZeroException;
import Exceptions.ValueNotFoundException;
import interfaces.*;
import model.ProgramState;

/**
 * Created by Lucian on 11/21/2015.
 */
public class HeapAllocation implements IStatement {

    /**
     * The varoiable name
     */
    private String variableName;

    private Expression expression;

    /**
     * Private contructor, use builder
     */
    public HeapAllocation(String variableName, Expression expression) {
        this.variableName = variableName;
//        this.expressionResult = expression.eval(symbolicTable);
        this.expression = expression;
//        pointer = heap.size();
//        heap.put(heap.size() ,expressionResult);
//        symbolicTable.put(variableName, pointer);
    }


    public String getVariableName() {
        return variableName;
    }

    public Expression getExpression() {
        return expression;
    }

    @Override
    public String toString() {
        return "HeapAllocation{ " + variableName + " -> " + expression.toString() + " }";
    }

    @Override
    public ProgramState execute(ProgramState programState) throws DivideByZeroException, ValueNotFoundException {
        IHeap<Integer, Integer> heap = programState.getHeap();
        IDictionary<String, Integer> myDictionary = programState.getMyDictionary();
        int pointer = heap.size();
        heap.put(pointer, getExpression().eval(myDictionary, heap));
        myDictionary.put(getVariableName(), pointer);
        return null;
    }
}

File: C#/ToyLanguage/ToyLanguage/Model/Statements/HeapAllocation.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using ToyLanguage.Interfaces;

namespace ToyLanguage.Model.Statements
{
    class HeapAllocation : IMyStatement
    {

        /**
         * The varoiable name
         */
        private String variableName;

        private IExpressions expression;

        /**
         * Private contructor, use builder
         */
        public HeapAllocation(String variableName, IExpressions expression)
        {
            this.variableName = variableName;
            //        this.expressionResult = expression.eval(symbolicTable);
            this.expression = expression;
            //        pointer = heap.size();
            //        heap.put(heap.size() ,expressionResult);
            //        symbolicTable.put(variableName, pointer);
        }


        public String getVariableName()
        {
            return variableName;
        }

        public IExpressions getExpression()
        {
            return expression;
        }


        public string MyToString()
        {
            return "HeapAllocation{ " + variableName + " -> " + expression.MyToString() + " }";
        }

        public ProgramState execute(ProgramState programState)
        {
            int pointer = programState.getHeap().size() + 1;
            programState.getHeap().put(pointer,expression.eval(programState.getMyDictionary(), programState.getHeap()));
            programState.getMyDictionary().put(variableName, pointer);
            return null;
        }
    }
}
